lib/Serialization/Serialization.h

```diff
@@ -120,29 +84,140 @@ class Serializer : public SerializerBase {
+  /// non-identifiers).
+  std::vector<StringRef> StringsToWrite;
+
+  /// The last assigned IdentifierID for uniqued strings from this module.
+  ///
+  /// Note that special module IDs and IDs of special names must not be valid
+  /// IdentifierIDs, except that 0 will always represent the empty identifier.
+  uint32_t /*IdentifierID*/ LastUniquedStringID =
+      serialization::NUM_SPECIAL_IDS - 1;
+
+  /// Helper for serializing entities in the AST block object graph.
+  ///
+  /// Keeps track of assigning IDs to newly-seen entities, and collecting
+  /// offsets when those entities are ready to be serialized.
+  ///
+  /// \tparam T The AST type that's being serialized
+  /// \tparam ID The ID type for encoding into the bitstream
+  /// \tparam RecordCode_ The code for the offsets record in the Index block for
+  /// referring to these entities.
+  template <typename T, typename ID, unsigned RecordCode_>
+  class Serialized {
+    /// Map of entities to assigned ID numbers.
+    llvm::DenseMap<T, ID> IDs;
+
+    /// All entities that still need to be written.
+    ///
+    /// This is a queue and not simply a vector because serializing one
+    /// entity might result in another one getting queued up.
+    std::queue<T> EntitiesToWrite;
+
+    /// The offsets of entities that have already been written.
+    ///
+    /// `Offsets[entityID - 1]` gives the offset for a particular entity. (0 is
+    /// reserved for "empty" entities.)
+    std::vector<BitOffset> Offsets;
+
+    /// The most recently assigned sequential ID.
+    ///
+    /// 0 is used for "empty" entities, so by starting at 0 we ensure that the
+    /// first entity is assigned 1 as its ID.
+    uint32_t LastID = 0;
+  public:
+    /// The code for the offsets record in the Index block for referring to
+    /// these entities.
+    static const unsigned RecordCode = RecordCode_;
+
+    /// Assigns \p entity an ID and schedules it for writing.
+    ///
+    /// If \p entity has been seen before, returns the existing ID.
+    ///
+    /// 0 is used for "empty" entities (those that coerce to \c false ).
+    ID addRef(T entity) {
+      assert(ID() == 0 && "bad default constructor for ID");
+      if (!entity)
+        return 0;
+
+      auto &entityID = IDs[entity];
+      if (entityID == ID()) {
+        EntitiesToWrite.push(entity);
+        entityID = ++LastID;
+      }
+      return entityID;
+    }
+
+    /// Returns true if \p entity has been seen before (i.e. ever passed to
+    /// #addRef).
+    bool hasRef(T entity) const {
+      return IDs.find(entity) != IDs.end();
+    }
+
+    bool hasMoreToSerialize() const {
+      return !EntitiesToWrite.empty();
+    }
+
+    /// Returns the next entity to be written.
+    T peekNext() const {
+      return EntitiesToWrite.front();
+    }
+
+    /// Records that the next entity will be written at \p offset, and returns
+    /// it so it can be written.
+    T popNext(BitOffset offset) {
+      T result = EntitiesToWrite.front();
```

Comment:
Same as above -- potential crash if someone forgot to check `hasMoreToSerialize()`.

Reply:
Ah, I see -- this function is only being called in a particular context where you check the condition so perhaps this is fine...

Reply:
I could make it an explicit assertion instead of relying on the STL to catch it. Would that help?

Reply:
Returning an optional (1) is safe, (2) let's you shorten the while loops:
```
    while (DeclsToSerialize.hasMoreToSerialize()) {
      auto next = DeclsToSerialize.popNext(Out.GetCurrentBitNo());
      writeDecl(next);
    }
```
becomes
```
    while (auto next = DeclsToSerialize.popNext(Out.getCurrentBitNo()))
      writeDecl(next.getValue());
```
and (3) let's you put an assertion when forcibly popping, so
```
    (void)GenericEnvironmentsToSerialize.popNext((genericSigID << 1) | 0x01);
```
becomes
```
    assert(GenericEnvironmentsToSerialize.popNext((genericSigID << 1) | 0x01).hasValue() && "[add justification here]");
```

Seems like a win-win -- what do you think?

Reply:
:-/ Take a closer look at writeNextGenericEnvironment. (I wouldn't have added `peek` at all if not for that.

Reply:
Ah, I see, you don't propose to get rid of `hasMoreToSerialize`. All right.